File: core/examples_cpp/tensors_0.cpp
#include <iostream>
#include "core/core_messages/core_messages.h"
#include "core/core_nodes/frequency_generator.h"
#include "core/core_nodes/map_fun.h"
#include "core/core_nodes/callback_fun.h"
#include "core/core_nodes/message_printer.h"
#include "core/util/utils.h"

using namespace roboflex::core;
using namespace roboflex::nodes;

int main() {
    FrequencyGenerator frequency_generator(2.0);

    auto tensor_creator = MapFun([](MessagePtr m) {
        xt::xtensor<double, 2> d = xt::ones<double>({2, 3}) * m->message_counter();
        return TensorMessage<double, 2>::Ptr(d);
    });

    auto message_printer = MessagePrinter("MESSAGE IS:");

    auto tensor_printer = CallbackFun([](MessagePtr m) {
        xt::xtensor<double, 2> d = TensorMessage<double, 2>(*m).value();
        std::cout << "TENSOR IS:" << std::endl << d << std::endl;
    });

    frequency_generator > tensor_creator > message_printer > tensor_printer;
    frequency_generator.start();
    sleep_ms(3000);
    frequency_generator.stop();

    std::cout << "DONE" << std::endl;
    return 0;
}

File: roboflex/core/examples/python/tensors_0.py
import time
import numpy as np
from roboflex.core import Node, FrequencyGenerator, MapFun, MessagePrinter, CallbackFun


# This example shows how to create a graph of nodes that pass messages containing
# numpy tensors (that can be interpreted at the c++-level as xtensor or eigen 
# objects) between each other in a chain.


# ----------- 
# create nodes of the graph

# The first node will signal at 2 hz. FrequencyGenerator is a library node that
# simply signals a BlankMessage at a given frequency. It runs in a thread, and must
# be started and stopped.
frequency_generator = FrequencyGenerator(2.0)

# Next, this node creates a message containing a numpy tensor. This python dict 
# will be encapsulated into a DynoFlex message, and be passed to the next node in 
# the graph.
tensor_creator = MapFun(lambda m: {"t": np.ones((2, 3)) * m.message_counter})

# These nodes print stuff out.
message_printer = MessagePrinter("MESSAGE IS:")

# Node is designed for subclassing.
class MyTensorPrinter(Node):
    def receive(self, m):
        print("MY TENSOR IS:", type(m["t"]), m["t"].shape, m["t"].dtype, "\n", m["t"])

tensor_printer = MyTensorPrinter()

# ----------- 
# connect nodes of the graph. It's easy to distribute the graph into
# multiple cpus using nodes in roboflex/transport.
frequency_generator > tensor_creator > message_printer > tensor_printer

# ----------- 
# start the root node (the other nodes, in this case, will run in the root node's thread).
frequency_generator.start()

# ----------- 
# go for a while
time.sleep(3)
    
# ----------- 
# stop the root node
frequency_generator.stop()

print("DONE")
